File: client/features/utility_measure.cc
#include "utility_measure.h"

#include "client/client.h"
#include "util/debug_util.h"

namespace tetris {

UtilityMeasure::UtilityMeasure() :
    _logger{debug::Logger::get()}
{}

UtilityMeasure::~UtilityMeasure()
{}

FeatureID UtilityMeasure::handshake()
{
    tetris::ClientMessage msg;

    msg.set_type(tetris::ClientMessage::FEATURE_SUBSCRIBE);
    auto feature_info = msg.mutable_feature_info();
    feature_info->set_type(tetris::ClientMessage::FeatureInfo::UTILITY_MEASURE);

    auto response = this->get_client()->send(msg);
    if ((response.type() == tetris::ServerResponse::FEATURE_ACKNOWLEDGE) && response.has_feature_ack_info()) {
        return response.feature_ack_info().id();
    }

    return -1;
}

ClientResponse UtilityMeasure::handle(const ServerMessage &msg)
{
    tetris::ClientResponse response{};

    if (_utility_measures.size() > 0) {
        response.set_type(ClientResponse::UTILITY_UPDATE);
    
         response.set_utility(_utility_measures.back());
    } else {
        response.set_type(ClientResponse::UTILITY_RETRY);
    }

    return response;
}

void UtilityMeasure::update_utility(float new_measure)
{
    _utility_measures.push_back(new_measure);
}

void UtilityMeasure::clear_utility()
{
    _utility_measures.clear();
}

};

File: client_py/pytetrisclient/features/utilityMeasure.py
import logging

from proto.tetris_pb2 import ClientMessage, ServerResponse, ClientResponse, ServerMessage, FeatureInfo

from pytetrisclient.client import Client
from pytetrisclient.utils.protobufUtil import ProtobufUtil
from pytetrisclient.feature import Feature


class UtilityMeasure(Feature):
    def __init__(self):
        super()._init__()

        self._utility_measures = list()

    def handle(self, msg: ServerMessage):
        response = ClientResponse()

        if len(self._utility_measures) > 0:
            response.type = ClientResponse.UTILITY_UPDATE
            response.utility = sum(self._utility_measures)/len(self._utility_measures)
            self._utility_measures.clear()
        else:
            response.type = ClientResponse.UTILITY_RETRY

        return response

    def need_handshake(self):
        return True

    def handshake(self):
        msg = ClientMessage()
        msg.type = ClientMessage.FEATURE_SUBSCRIBE

        feature_info = FeatureInfo()
        feature_info.type = FeatureInfo.UTILITY_MEASURE
        msg.feature_info = feature_info

        response = self._client.send(msg)

        if response.type == ServerResponse.FEATURE_ACKNOWLEDGE:
            return response.feature_ack_info.id

        raise RuntimeError("Could not register UtilityMeasure feature with the server")

    def update_utility(self, new_measure: float):
        self._utility_measures.append(new_measure)

    def clear_utility(self):
        self._utility_measures.clear()
